### PR TITLE
Improve softmax: type checks, stability, axis support, tests, docs

### LINKED ISSUE #13171: Softmax function missing type checking and other enhancements
### Feature description

The softmax function does not have specification for the input type required to compute softmax. At the least, the function argument should have strict type checking, in place, to prevent crashing on non-supported inputs such as string or dict.

Other recommendations are:

1. Numerical stability: subtract the max on the reduction axis before exponentiating to avoid overflow.
2. Make the function N‑D and axis-aware: add an axis parameter (default -1) so it works for 1D/2D/3D tensors and is explicit about where probabilities sum to 1.
3. Input normalization and dtype: convert inputs with np.asarray(..., dtype=float) so lists/tuples/ints behave correctly.
4. Use keepdims=True on reductions so broadcasting keeps shapes correct.
5. Add simple type hints and a clearer docstring describing inputs/outputs and axis semantics.
6. Add unit tests (happy path + edge cases: scalar, single element, identical logits, large values, empty) and a small __main__ demo.
7. Defensive checks: raise meaningful error for empty input and document behavior for NaNs/Infs.

A sample improved version is as follows:

```
import numpy as np
from typing import Optional

def softmax(vector, axis: Optional[int] = -1) -> np.ndarray:
    """
    Compute the softmax of `vector` along `axis` in a numerically-stable way.

    Parameters
    ----------
    vector : array_like
        Input data (vector, matrix, tensor). Will be converted to float ndarray.
    axis : int or None, optional
        Axis along which to compute softmax. If None, compute softmax over
        the flattened array (single distribution). Default is -1 (last axis).

    Returns
    -------
    np.ndarray
        Same shape as `vector`, with softmax applied along `axis`. Probabilities sum to 1
        along `axis` (or to 1 overall if axis is None).

    Raises
    ------
    ValueError
        If input is empty.
    """
    try:
        x = np.asarray(x, dtype=float)
    except Exception as e:
        raise ValueError(f"Could not convert input to float ndarray: {e}")
    
    if vector.size == 0:
        raise ValueError("softmax input must be non-empty")

    if axis is None:
        # flatten to single distribution
        vector_max = np.max(vector)
        e_vector = np.exp(vector - vector_max)
        return e_vector / e_vector.sum()

    # subtract max along axis with keepdims for numerical stability/broadcasting
    vector_max = np.max(vector, axis=axis, keepdims=True)
    e_vector = np.exp(vector - vector_max)
    denom = e_vector.sum(axis=axis, keepdims=True)
    return e_vector / denom

if __name__ == "__main__":
    print(softmax((0,)))

```
credit to Github Copilot

### PR DESCRIPTION
**Pull request by shreyasNaik0101**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #13171".

### CODE REVIEW

File: maths/softmax.py [hunk 1]
```diff
@@ -1,56 +1,96 @@
...
-    Returns:
-        softmax_vec (np.array): The input numpy array  after applying
-        softmax.
+    # Typical 1D input
+    result = softmax([1, 2, 3])
+    npt.assert_almost_equal(result.sum(), 1)
 
-    The softmax vector adds up to one. We need to ceil to mitigate for
-    precision
-    >>> float(np.ceil(np.sum(softmax([1,2,3,4]))))
```
Comment (cclauss) [inline maths/softmax.py:31]: Removing doctests goes against the recommendations in CONTRIBUTING.md.
